src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,108 @@
+            _inputFileElement = inputFileElement;
+            _maxChunkSize = maxChunkSize;
+            _chunkSequence = new PreFetchingSequence(
+                FetchEncodedChunk,
+                (File.Size + _maxChunkSize - 1) / _maxChunkSize,
+                Math.Max(1, maxBufferSize / _maxChunkSize)); // Degree of parallelism on fetch.
+            _currentChunkDecodingBuffer = new byte[_maxChunkSize];
+        }
+
+        protected override async ValueTask<int> CopyFileDataIntoBuffer(long sourceOffset, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            var totalBytesCopied = 0;
+
+            while (destination.Length > 0)
+            {
+                // If we don't yet have a chunk, or it's fully consumed, get the next one.
+                if (!_currentChunk.HasValue || _currentChunkDecodingBufferConsumedLength == _currentChunk.Value.LengthBytes)
+                {
+                    // If we've already read some data, and the next chunk is still pending,
+                    // then just return now rather than awaiting.
+                    if (totalBytesCopied > 0 && _chunkSequence.TryPeekNext(out var nextChunk) && !nextChunk.Base64.IsCompleted)
+                    {
+                        break;
+                    }
+
+                    _currentChunk = _chunkSequence.ReadNext(cancellationToken);
+
+                    var currentEncodedChunk = await _currentChunk.Value.Base64;
+
+                    DecodeChunkToBuffer(currentEncodedChunk, _currentChunkDecodingBuffer, 0, _currentChunk.Value.LengthBytes);
+
+                    _currentChunkDecodingBufferConsumedLength = 0;
+                }
+
+                // How much of the current chunk can we fit into the destination?
+                var numUnconsumedBytesInChunk = _currentChunk.Value.LengthBytes - _currentChunkDecodingBufferConsumedLength;
+                var numBytesToTransfer = Math.Min(numUnconsumedBytesInChunk, destination.Length);
+
+                if (numBytesToTransfer == 0)
+                {
+                    break;
+                }
+
+                // Perform the copy.
+                new Memory<byte>(_currentChunkDecodingBuffer, _currentChunkDecodingBufferConsumedLength, numBytesToTransfer).CopyTo(destination);
+                destination = destination.Slice(numBytesToTransfer);
+                _currentChunkDecodingBufferConsumedLength += numBytesToTransfer;
+                totalBytesCopied += numBytesToTransfer;
+            }
+
+            return totalBytesCopied;
+        }
+
+        private EncodedFileChunk FetchEncodedChunk(long index, CancellationToken cancellationToken)
+        {
+            var sourceOffset = index * _maxChunkSize;
+            var chunkLength = (int)Math.Min(_maxChunkSize, File.Size - sourceOffset);
+            var task = _jsRuntime.InvokeAsync<string>(
+                InputFileInterop.ReadFileData,
+                cancellationToken,
```

Comment:
From what I understand, this cancellation token is passed by the user and is "for the whole file". I think we should have a separate "per chunk" cancellation with a default value. Otherwise its really easy to provide an unsafe value (CancellationToken.None) or a very high value if I'm trying to upload a large file that allows individual chunks to be awaited for a long time.

Reply:
I've changed how/when data is fetched - please let me know if you have the same concern.